import os
import pandas as pd

# configfile: 'test.yaml'
# configfile: 'config_local.yaml' # local wgs using paired-end data
# configfile: 'config_local_se.yaml' # local wgs using single-end data
configfile: 'config_local_se_pop.yaml' # local wgs using single-end data
# configfile: 'exp_sim_local.yaml'
# configfile: "config.yaml"
# configfile: "config_mouse.yaml"

''' Load from config '''
CHROM = config['CHROM']
INDIV = config['INDIV']
EXP_LABEL = config['EXP_LABEL']
ALN_MODE = config['ALN_MODE']
assert ALN_MODE in ['single-end', 'paired-end']
READS1 = config['READS1']
READS2 = config['READS2']
GROUP = config['GROUP']
POP_LEVEL = config['POP_LEVEL']
ALN_MAPQ_THRSD = config['ALN_MAPQ_THRSD']
POP_THRSD = config['POP_THRSD']
POP_STOCHASTIC = config['POP_STOCHASTIC']
POP_BLOCK_SIZE = config['POP_BLOCK_SIZE']
POP_USE_LD = config['POP_USE_LD']

USE_PREBUILT = config['USE_PREBUILT']
SORT_SAM = config['SORT_SAM']

DIR = config['DIR']
GENOME = config['GENOME']
DIR_VCF = config['DIR_VCF']
VCF_MAP_CHROM = config['VCF_MAP_CHROM']
VCF_PREFIX = config['VCF_PREFIX']
VCF_SUFFIX = config['VCF_SUFFIX']
CHR_PREFIX = config['CHR_PREFIX']
LENGTH_MAP = config['LENGTH_MAP']
CHROM_MAP = config['CHROM_MAP']

FAMILY = config['FAMILY']
SPOP = config['SPOP']
BCFTOOLS = config['BCFTOOLS']
SAMTOOLS = config['SAMTOOLS']
LEVIOSAM = config['LEVIOSAM']
PYTHON = config['PYTHON']
DIR_SCRIPTS = config['DIR_SCRIPTS']

THREADS = config['THREADS']
RAND_SEED = config['RAND_SEED']

# Experiments
EXP = True
# EXP = False

#### Below args are only used in experiments
if EXP:
    IF_REAL_DATA = config['IF_REAL_DATA']
    BIAS_MIN_READ_COUNT = config['BIAS_MIN_READ_COUNT']
    BIAS_TAIL_THRDS = config['BIAS_TAIL_THRDS']
    BEDTOOLS = config['BEDTOOLS']
    DIR_SCRIPTS_EXP = config['DIR_SCRIPTS_EXP']
    GRAPH_AF_THRSD = config['GRAPH_AF_THRSD']
    TIME = config['TIME']
    VG = config['VG']
    DBSNP_COMMON = config['DBSNP_COMMON']
    GATK = config['GATK']
    PICARD = config['PICARD']
''''''

# Bowtie 2 index extensions
IDX_ITEMS = ['1', '2', '3', '4', 'rev.1', 'rev.2']

# Prefixes and directory paths for major-allele reference contruction and indexing
PREFIX_MAJOR_F = os.path.join(DIR, 'major/{CHROM}_filtered_major')
PREFIX_MAJOR = os.path.join(DIR, 'major/chr{CHROM}_maj')
DIR_MAJOR = os.path.join(DIR, 'major')

# Prefixes and directory paths for population reference contruction and indexing
DIR_POP_GENOME = os.path.join(DIR, 'pop_genome/')
POP_DIRNAME = 'thrds{0}_S{1}_b{2}_ld{3}'.format(POP_THRSD, POP_STOCHASTIC, POP_BLOCK_SIZE, POP_USE_LD)
WG_POP_GENOME_SUFFIX = EXP_LABEL + '-' + POP_LEVEL + '_{GROUP}_' + POP_DIRNAME
DIR_POP_GENOME_BLOCK = os.path.join(DIR_POP_GENOME, POP_DIRNAME + '/')
DIR_POP_GENOME_BLOCK_IDX = os.path.join(DIR_POP_GENOME_BLOCK, 'indexes/')

# Prefix and directory paths for experiments
DIR_FIRST_PASS = os.path.join(DIR, 'experiments/{INDIV}/')
DIR_SECOND_PASS = os.path.join(DIR, 'experiments/{INDIV}/' + POP_DIRNAME)
PREFIX_SECOND_PASS = os.path.join(DIR_SECOND_PASS, EXP_LABEL + '-major-' + ALN_MAPQ_THRSD + '-{GROUP}-' + POP_DIRNAME)

# Bias results directory
DIR_RESULTS = os.path.join(DIR, 'results')
DIR_RESULTS_BIAS = os.path.join(DIR, 'results/bias')
DIR_VG = os.path.join(DIR, 'vg/')

''' Snakemake modules '''
if EXP:
    DIR_PER = os.path.join(DIR, 'personalized/{INDIV}/')
    PREFIX_PER = os.path.join(DIR_PER, EXP_LABEL)
    include: 'experiments/prepare_per_and_grc.Snakefile'
    include: 'experiments/graph_index.Snakefile'
    if not IF_REAL_DATA:
        READS1 = PREFIX_PER + '_1.fq'
        READS2 = PREFIX_PER + '_2.fq'
    if ALN_MODE == 'single-end':
        include: 'experiments/exp_alignment_single_end.Snakefile'
        include: 'experiments/graph_alignment_single_end.Snakefile'
    else:
        include: 'experiments/exp_alignment_paired_end.Snakefile'
        include: 'experiments/graph_alignment_paired_end.Snakefile'
    include: 'experiments/evaluate_bias.Snakefile'
    if not IF_REAL_DATA:
        include: 'experiments/calc_mapping_acc.Snakefile'
    include: 'experiments/exp_shared.Snakefile'
    include: 'experiments/variant_calling.Snakefile'

# Functions
include: '../reference_flow/snakemake/shared/functions.Snakefile'

if not USE_PREBUILT:
    # Prepare pop genome and indexes
    #      check: 'prepare_pop_genome.done'
    include: '../reference_flow/snakemake/shared/prepare_pop_genome.Snakefile'
    
    # Prepare grc and major genome and indexes
    #      check: 'prepare_standard_genome.done'
    include: '../reference_flow/snakemake/shared/prepare_standard_genome.Snakefile'

# Align reads to grch major, reference flow, personalized genomes
#      check: 'alignment_refflow.done'
if ALN_MODE == 'single-end':
    include: '../reference_flow/snakemake/shared/alignment_single_end.Snakefile'
elif ALN_MODE == 'paired-end':
    include: '../reference_flow/snakemake/shared/alignment_paired_end.Snakefile'

# Lift and sort reads
#      check: 'leviosam.done', 'sort.done'
include: '../reference_flow/snakemake/shared/lift_and_sort.Snakefile'

# TODO_LIST = ['alignment_refflow.done', 'leviosam.done']
TODO_LIST = ['alignment_refflow.done']
if not USE_PREBUILT:
    TODO_LIST.append('prepare_pop_genome.done')
    TODO_LIST.append('prepare_standard_genome.done')
if SORT_SAM:
    TODO_LIST.append('sort.done')

if EXP:
#     TODO_LIST.append('prepare_per.done')
#     TODO_LIST.append('bias_exp.done')
#     TODO_LIST.append('alignment_grc_per.done')
#     vg
#     TODO_LIST.append('vg_index.done')
    TODO_LIST.append('vg_map.done')
#     TODO_LIST.append(os.path.join(DIR_RESULTS, 'all.tsv'))
    TODO_LIST.append('bias_exp.done')
#     TODO_LIST.append('ht_align.done')
#    TODO_LIST.append('var_calling.done')
    if not IF_REAL_DATA:
        TODO_LIST.append('accuracy.done')

''' Snakemake rules '''
rule all:
    input:
        expand(os.path.join(DIR, '{task}'), task = TODO_LIST)

rule filter_vcf:
    input:
        vcf = os.path.join(DIR_VCF, VCF_PREFIX + '{CHROM}' + VCF_SUFFIX),
        chrom_map = CHROM_MAP
    output:
        vcf = temp(os.path.join(DIR, '{CHROM}_filtered.vcf'))
    run:
        # Take PASS variants
        # Does not remove mnps, since they will be needed for constructing personalized reference genome, 
        # and will be removed when building major and refflow references.
        if VCF_MAP_CHROM:
            shell('{BCFTOOLS} view -r {wildcards.CHROM} -c 1 -f PASS {input.vcf} | {BCFTOOLS} annotate --rename-chrs {input.chrom_map} -o {output.vcf}')
        else:
            shell('{BCFTOOLS} view -r chr{wildcards.CHROM} -c 1 -f PASS {input.vcf} -o {output.vcf}')

rule aggregate_vcf:
    input:
        vcf = expand(os.path.join(DIR, '{CHROM}_filtered.vcf'), CHROM = CHROM)
    output:
        vcf = os.path.join(DIR, EXP_LABEL + '_filtered.vcf.gz')
    shell:
        '{BCFTOOLS} concat -O z -o {output.vcf} {input.vcf}'

'''Get variants above an allele frequency threshold.

This is needed to build indexes for graph mappers such as vg and HISAT2.

Rules:
    filter_vcf_by_freq
    build_vcf_index
'''
rule filter_vcf_by_freq:
    input:
        vcf = os.path.join(DIR, EXP_LABEL + '_filtered.vcf')
    output:
        vcfgz = os.path.join(DIR, EXP_LABEL + '-{}.vcf.gz'.format(GRAPH_AF_THRSD))
    shell:
        '{BCFTOOLS} view -O z -m2 -M2 -V mnps,other -q {GRAPH_AF_THRSD} {input.vcf} > {output.vcfgz};'

rule build_vcf_index:
    input:
        vcfgz = os.path.join(DIR, EXP_LABEL + '-{}.vcf.gz'.format(GRAPH_AF_THRSD))
    output:
        csi = os.path.join(DIR, EXP_LABEL + '-{}.vcf.gz.csi'.format(GRAPH_AF_THRSD)),
        tbi = os.path.join(DIR, EXP_LABEL + '-{}.vcf.gz.tbi'.format(GRAPH_AF_THRSD))
    shell:
        'tabix -p vcf {input.vcfgz};'
        '{BCFTOOLS} index {input.vcfgz};'

